File: leet-code/461. Hamming Distance/HammingDistance461_2.java
package com.example;

public class HammingDistance461_2 {
    public int hammingDistance(int x, int y) {
        return countOneBit(toBinary(x ^ y));
    }

    private String toBinary(int num) {
        StringBuilder sb = new StringBuilder();
        while (num > 0) {
            sb.insert(0, num % 2);
            num /= 2;
        }
        return sb.toString();
    }

    private int countOneBit(String binary) {
        int count = 0;
        for (char b : binary.toCharArray()) {
            if (b == '1') {
                count += 1;
            }
        }
        return count;
    }
}

File: leet-code/461. Hamming Distance/HammingDistance461_2.kt
package com.example

import kotlin.math.abs
import kotlin.math.max

class HammingDistance461_2 {
    fun hammingDistance(x: Int, y: Int): Int {
        val binary1 = convertToBinaryString(x)
        val binary2 = convertToBinaryString(y)
        val maxLength = max(binary1.length, binary2.length)
        val paddedBinary1 = padLeft(binary1, maxLength)
        val paddedBinary2 = padLeft(binary2, maxLength)
        return IntRange(0, maxLength - 1)
            .map { if (paddedBinary1[it] != paddedBinary2[it]) 1 else 0 }
            .filter { it == 1 }
            .size
    }

    private fun convertToBinaryString(num: Int): String {
        var number = num
        val sb = StringBuilder()
        while (number > 0) {
            sb.insert(0, number % 2)
            number /= 2
        }
        return sb.toString()
    }

    private fun padLeft(binary: String, length: Int): String {
        val binaryLength = binary.length
        val diff = abs(binaryLength - length)
        val sb = StringBuilder(binary)
        for (i in 0 until diff) {
            sb.insert(0, "0")
        }
        return sb.toString()
    }
}
